### DISCUSSION TITLE
Building Electron app for Windows on Linux

### DISCUSSION THREAD

**Discussion by ccheraa** [eclipse-theia/theia#9078] [Q&A] [answered]
I am trying to build my Theia app for windows on `electronuserland/builder:wine` docker image, I have prebuilt binaries for windows for the native modules hosted on a server, the build works without problems.
While building it mentions that it uses `drivelist` prebuilt binaries, and rebuilds all the other native modules (`native-ketymap`, `nsfw`, `oniguruma`, etc).

But when I try to run the App on Windows, it says:
```
Error: \\?\path\to\app\resources\app\node_modules\nsfw\build\Release\nsfw.node is not a valid Win32 application.
```
I replaced the file with my the binary previously built on a Windows machine, so I got the same error for `find-git-repositories\build\Release\findGitRepos.node`

And after replaceing `findGitRepos.node` with the prebuilt one, it worked.

I tried with [github.com/theia-ide/theia-apps/tree/master/theia-electron](https://github.com/theia-ide/theia-apps/tree/master/theia-electron) and I got the same problem.

I am not sure if this is a problem with `Theia` or `electron-builder`.

Is there a way to force electron-builder to download and use the prebuilt binaries of all the native modules instead of downloading `drivelist` only?

This is my `package.json` file:

```json
{
  "name": "theia-test",
  "version": "1.0.0",
  "license": "MIT",
  "private": true,
  "main": "scripts/theia-electron-main.js",
  "scripts": {
    "prepare": "yarn run clean && yarn build",
    "clean": "theia clean",
    "build": "theia build --mode development",
    "start": "theia start",
    "package": "yarn run build && electron-builder",
    "package:all": "yarn run build && electron-builder -lmw"
  },
  "theia": {
    "target": "electron",
    "frontend": {
      "config": {
        "applicationName": "Theia test"
      }
    }
  },
  "dependencies": {
    "@theia/callhierarchy": "^1.9.0",
    "@theia/electron": "^1.9.0",
    "@theia/file-search": "^1.9.0",
    "@theia/git": "^1.9.0",
    "@theia/markers": "^1.9.0",
    "@theia/messages": "^1.9.0",
    "@theia/mini-browser": "^1.9.0",
    "@theia/monaco": "^1.9.0",
    "@theia/navigator": "^1.9.0",
    "@theia/outline-view": "^1.9.0",
    "@theia/plugin-ext-vscode": "^1.9.0",
    "@theia/preferences": "^1.9.0",
    "@theia/preview": "^1.9.0",
    "@theia/search-in-workspace": "^1.9.0",
    "@theia/terminal": "^1.9.0",
    "fs-extra": "^9.0.1",
    "xterm": "^4.9.0",
    "yargs": "^16.2.0"
  },
  "devDependencies": {
    "@theia/cli": "^1.9.0",
    "electron-builder": "^22.9.1"
  }
}
```

**Comment by kittaakos** [comment, marked as answer]
> windows on `electronuserland/builder:wine`

I ran into the same problem a long time ago. You cannot use that docker image with `wine` to build electron apps for Windows. You really need a Windows machine. See here: https://github.com/theia-ide/theia-apps/tree/master/theia-electron#troubleshooting :/

I think the problem is/was, that even if you have `wine` installed, Node.js (`electron-builder`) will see `'linux'` platform as `os.platform()` instead of `'win32'`' and will build you Linux natives. They won't run on Windows.

**Reply by kittaakos** [reply]
Maybe you can try to force the Windows platform with `-w` (or `-c.platform win32` if it works at all). The `electron-builder` community for sure knows why it does not work.

**Reply by ccheraa** [reply]
That does not work.

After digging around, I found a way to force `electron-builder` to download the prebuilt binaries instead of rebuilding them, it's more of a workaround though.

All you have to do is add `prebuild-install` to the dependencies in `package.json` of the module you want to download the prebuilt binaries for.

I also opened an [issue (feature request)](https://github.com/develar/app-builder/issues/49) to allow us to do this in a less ugly way.
